### PR TITLE
Fix topic rewrite wildcard

### PR DESCRIPTION
**Pull request by JimMoen**
**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/emqx/emqx/blob/master/CONTRIBUTING.md.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/` dir
- [x] For EMQX 4.x: `appup` files updated (execute `scripts/update-appup.sh emqx`)
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up
- [ ] In case of non-backward compatible changes, reviewer should check this item as a write-off, and add details in **Backward Compatibility** section

### CODE REVIEW

**Comment by terry-xiaoyu** [comment]
Rebase to main-v4.3 and release it in e4.4.12

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 1]
```diff
@@ -75,30 +74,60 @@ description() ->
...
+validate_rules([Rule = {_, Type, _, _, _} | RestRules], {PubRules, SubRules}) ->
+    case {Type, validate_rule(Rule)} of
+        {pub, {ok, NRule}} ->
+            validate_rules(RestRules, {[NRule| PubRules], SubRules});
+        {sub, {ok, NRule}} ->
+            validate_rules(RestRules, {PubRules, [NRule | SubRules]});
+        {_, {error, {Direction, T}, Reason}} ->
+            ?LOG(alert, "Topic rewrite ~p rule[~p] ~p topic ~p validate error, Reason: ~p",
+                 [assert_type(Type), rule_index(Type, PubRules, SubRules), Direction, T, Reason]),
+            {[], []}
```
Comment (terry-xiaoyu) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:107]: We'd better only discard the bad rule, not all of the rules if only one of them is invalid.

File: changes/v4.3.22-en.md [hunk 2]
```diff
@@ -45,6 +45,8 @@
 - For Rule-Engine resource creation failure, delay before the first retry [#9313](https://github.com/emqx/emqx/pull/9313).
   Prior to this change, the retry delay was added *after* the retry failure.
 
+- Added topic validation for `emqx_mod_rewrite`. The dest topics contains wildcards are not allowed to publish [#9359](https://github.com/emqx/emqx/issues/9359).
```
Comment (zmstone) [inline changes/v4.3.22-en.md:48]: missed the v4.3.22 train.
please create a new release. v4.3.23

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 3]
```diff
@@ -75,30 +74,60 @@ description() ->
...
+        {error, Reason} ->
+            {error, {source, Source}, Reason}
+    end.
+
+log_rules(Type, Rules) ->
+    log_rules_(Type, Rules, 1).
 
-log_rule([], _Index) -> ok;
-log_rule([{Type, Topic, Re, Dest} | Rules], Index) ->
+log_rules_(_Type, [], _Index) -> ok;
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:126]:
```suggestion
do_log_rules(_Type, [], _Index) -> ok;
```

Comment (zmstone) [reply lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:126]: or `log_rules_loop`
if i remember correctly, elvis does not like functions ending with `_`

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 4]
```diff
@@ -138,3 +167,19 @@ filter_client_binds(Binds) ->
...
                    end,
                  Binds).
+
+assert_type(pub) -> 'PUBLISH';
+assert_type(sub) -> 'SUBSCRIBE'.
+
+dest_topic_type(pub) -> name;
+dest_topic_type(sub) -> filter.
+
+-type(topic() :: binary()).
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:177]: move to the head of the module.

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 5]
```diff
@@ -138,3 +167,19 @@ filter_client_binds(Binds) ->
...
+-spec(validate_topic(name | filter, topic()) -> topic()).
+validate_topic(Type, Topic) ->
+    try emqx_topic:validate(Type, Topic) of
+        true -> {ok, Topic}
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:181]:
```suggestion
    try 
        true = emqx_topic:validate(Type, Topic),
        {ok, Topic}
```

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 6]
```diff
@@ -75,30 +74,60 @@ description() ->
...
+            validate_rules(RestRules, {[NRule| PubRules], SubRules});
+        {sub, {ok, NRule}} ->
+            validate_rules(RestRules, {PubRules, [NRule | SubRules]});
+        {_, {error, {Direction, T}, Reason}} ->
+            ?LOG(warning, "Topic rewrite ~p rule item validate error. The ~p topic ~p invalid, Reason: ~p",
+                 [assert_type(Type), Direction, T, Reason]),
+            validate_rules(RestRules, {PubRules, SubRules})
+    end.
+
+validate_rule({rewrite, Type, Source, Re, Dest}) ->
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:110]:
```suggestion
validate_rule({rewrite, Type, RewriteFrom, Re, RewriteTo}) ->
```

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 7]
```diff
@@ -75,30 +74,60 @@ description() ->
...
+    {PubRules, SubRules};
+
+validate_rules([Rule = {_, Type, _, _, _} | RestRules], {PubRules, SubRules}) ->
+    case {Type, validate_rule(Rule)} of
+        {pub, {ok, NRule}} ->
+            validate_rules(RestRules, {[NRule| PubRules], SubRules});
+        {sub, {ok, NRule}} ->
+            validate_rules(RestRules, {PubRules, [NRule | SubRules]});
+        {_, {error, {Direction, T}, Reason}} ->
+            ?LOG(warning, "Topic rewrite ~p rule item validate error. The ~p topic ~p invalid, Reason: ~p",
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:105]:
```suggestion
            ?LOG(warning, "Invalid rewrite rule ~p for topic '~s' discarded. Reason: ~p",
```

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 8]
```diff
@@ -138,3 +167,19 @@ filter_client_binds(Binds) ->
                     (_) -> true
                    end,
                  Binds).
+
+assert_type(pub) -> 'PUBLISH';
+assert_type(sub) -> 'SUBSCRIBE'.
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:172]:
```suggestion
type_to_name(pub) -> 'PUBLISH';
type_to_name(sub) -> 'SUBSCRIBE'.
```

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 9]
```diff
@@ -75,30 +74,60 @@ description() ->
...
+    validate_rules(Rules, {[], []}).
+
+validate_rules([], {PubRules, SubRules}) ->
     ?LOG(info, "[Rewrite] Load: pub rules count ~p sub rules count ~p",
         [erlang:length(PubRules), erlang:length(SubRules)]),
-    log_rule(PubRules, 1),
-    log_rule(SubRules, 1).
+    log_rules(pub, PubRules),
+    log_rules(sub, SubRules),
+    {PubRules, SubRules};
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:96]: Double check: should we reverse or NOT ?
if not, add a comment here to explain why NOT.
```suggestion
    {lists:reverse(PubRules), lists:reverse(SubRules)};
```

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 10]
```diff
@@ -75,30 +74,60 @@ description() ->
...
+
+compile({PubRules, SubRules}) ->
+    CompileRE =
+        fun({rewrite, Source, Re, Dest}) ->
+                {ok, MP} = re:compile(Re),
+                {rewrite, Source, MP, Dest}
+        end,
+    {lists:map(CompileRE, PubRules), lists:map(CompileRE, SubRules)}.
+
+validate_rules(Rules) ->
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:88]: add a test case for compile result ordering regression.

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 11]
```diff
@@ -75,30 +74,60 @@ description() ->
...
-    PubRules = [ begin
-                     {ok, MP} = re:compile(Re),
-                     {rewrite, Topic, MP, Dest}
-                 end || {rewrite, pub, Topic, Re, Dest}<- Rules ],
-    SubRules = [ begin
-                     {ok, MP} = re:compile(Re),
-                     {rewrite, Topic, MP, Dest}
-                 end || {rewrite, sub, Topic, Re, Dest}<- Rules ],
-    {PubRules, SubRules}.
+    do_log_rules(Type, Rules, Index + 1).
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:130]: do a list:partition first, and then apply compile_pub and compile_sub on the partitioned result.

File: lib-ce/emqx_modules/src/emqx_mod_rewrite.erl [hunk 12]
```diff
@@ -75,30 +76,57 @@ description() ->
...
+                    log_invalid_rule(to, Type, RewriteTo, Reason),
+                    false
+            end;
+        {error, Reason} ->
+            log_invalid_rule(from, Type, RewriteFrom, Reason),
+            false
+    end.
+
+log_invalid_rule(Direction, Type, Topic, Reason) ->
+    ?LOG(warning, "Invalid rewrite ~p rule for rewrite ~p topic '~s' discarded. Reason: ~p",
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_rewrite.erl:119]:
```suggestion
    ?LOG(warning, "Invalid rewrite ~p rule for rewrite ~p topic '~ts' discarded. Reason: ~p",
```

**Review by terry-xiaoyu — APPROVED** [review]

**Pull request merged**